cpp/include/cudf/detail/copy.hpp

```diff
@@ -88,17 +88,45 @@ std::unique_ptr<column> shift(
   rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief The data that is stored as anonymous bytes in the `packed_columns` metadata
+ * field
+ *
+ * @ingroup copy_split
+ *
+ * The metadata field of the `packed_columns` struct is simply an array of these.
+ * This struct is exposed here because it is needed by both contiguous_split, pack
+ * and unpack;
+ */
+struct serialized_column {
+  data_type type;
+  size_type size;
+  size_type null_count;
+  int64_t data_offset;       // offset into contiguous data buffer, or -1 if column data is null
+  int64_t null_mask_offset;  // offset into contiguous data buffer, or -1 if column data is null
+  size_type num_children;
+};
+
 /**
  * @copydoc cudf::contiguous_split
  *
  * @param stream CUDA stream used for device memory operations and kernel launches.
- */
-std::vector<contiguous_split_result> contiguous_split(
+ **/
+std::vector<packed_table> contiguous_split(
   cudf::table_view const& input,
   std::vector<size_type> const& splits,
   rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @copydoc cudf::pack
+ *
+ * @param stream Optional CUDA stream on which to execute kernels
+ **/
+packed_columns pack(cudf::table_view const& input,
```

Comment:
While starting to integrate this functionality into the RAPIDS Spark Accelerator, I ran across a case where I have a contiguous table that was **not** built via `contiguous_split`.  It's a table that was constructed on the host during Spark's built-in shuffle and copied to the device as a single buffer.  We'd like to use libcudf's `packed_columns` metadata to reconstruct our spilled buffers loaded from hostmem and/or disk, so we need the ability to construct the libcudf opaque metadata for one of these tables.  Therefore I'd like to see an API that could take a `table_view`, a device buffer start address, and a device buffer length, and generate only the table metadata without copying the GPU data.

Without this new metadata-only API we'll have to call this `pack` API which will needlessly copy the GPU data.  If that's something that should be tackled in a followup issue, I'm happy to file it.

Reply:
It's a pretty straightforward thing to add to this PR.  I would just expose this function in pack.cpp and then have contiguous_split call that as a postprocess instead of inline with what it's doing now.  

This would also clean up something I don't love about how the code is currently arranged :  contiguous_split.cu is producing the metadata (inline in the step where it builds the output columns) but pack.cpp is consuming it.  It would be cleaner if pack.cpp both produced and consumed the data.

Reply:
Added `packed_metadata()`  

@jrhemstad  A good amount of shuffling of code happens with this update.  Mostly de-exposing stuff via copy.hpp and centralizing the code better in pack.cpp

Two things worthy of discussion:
- `pack_metadata()` right now returns a `std::vector<uint8_t>` which is a little raw.  Normally this buffer is hidden in the `packed_columns` struct.   Might be worth wrapping it in it's own little container class just to be a little more formal?

- I'm constructing a table_view here and passing it to both the pack_metadata function and also using it to initialize the `table` field of the returned `packed_table` struct.  Any concerns here that the compiler won't be smart enough to optimize this initialize-then-copy step away?

Reply:
> pack_metadata() right now returns a std::vector<uint8_t> which is a little raw. 

Agreed, I think hiding it behind something more opaque would be nice. Maybe something as simple as this:
```
struct metadata{
   void* data(){
      return data_.data();
   }
   std::size_t size(){
      return data_.size();
   }
private
   std::vector<char> data_;
};
```